reflect/src/main/java/dagger/reflect/ReflectiveMembersInjector.java

```diff
@@ -87,43 +87,43 @@
       }
 
       if (!fieldProviders.isEmpty() || !methodProviders.isEmpty()) {
-        // [@Inject] Fields and methods in superclasses are injected before those in subclasses.
-        // we are walking up (getSuperclass), but spec says @Inject should be walking down the hierarchy
-        hierarchyBindings.addFirst(new ClassBindings(fieldProviders, methodProviders));
+        // Per JSR 330, fields and methods in superclasses are injected before those in subclasses.
+        // Since we are traversing upward in the hierarchy each injector is thus prepended to the
+        // collection so regular iteration will honor this contract.
+        classInjectors.addFirst(new ClassInjector<>(fieldProviders, methodProviders));
       }
 
       target = target.getSuperclass();
     }
 
-    return new ReflectiveMembersInjector<>(hierarchyBindings);
+    return new ReflectiveMembersInjector<>(classInjectors);
   }
 
-  private final Iterable<ClassBindings> classBindings;
+  private final Iterable<ClassInjector<T>> classBindings;
 
-  private ReflectiveMembersInjector(Iterable<ClassBindings> classBindings) {
+  private ReflectiveMembersInjector(Iterable<ClassInjector<T>> classBindings) {
     this.classBindings = classBindings;
   }
 
   @Override public void injectMembers(T instance) {
-    for (ClassBindings classBinding : classBindings) {
+    for (ClassInjector<T> classBinding : classBindings) {
       classBinding.injectMembers(instance);
     }
   }
 
-  private static final class ClassBindings {
+  private static final class ClassInjector<T> implements MembersInjector<T> {
     final Map<Field, Provider<?>> fieldProviders;
     final Map<Method, Provider<?>[]> methodProviders;
 
-    ClassBindings(
+    ClassInjector(
         Map<Field, Provider<?>> fieldProviders,
         Map<Method, Provider<?>[]> methodProviders) {
       this.fieldProviders = fieldProviders;
       this.methodProviders = methodProviders;
     }
 
-    void injectMembers(Object instance) {
-      // [@Inject] Constructors are injected first, followed by fields, and then methods.
-      // Note: the constructor injection is in dagger.reflect.Binding.UnlinkedJustInTime.dependencies
+    @Override public void injectMembers(T instance) {
```

Comment:
ah, that's why it was `T`, the interface

Reply:
Yeah there's no real reason to do this, but it was effectively free so why not